### ISSUE TITLE
[feature request] Add option to return matched / unmatched / unexpected in `load_state_dict`

### ISSUE THREAD

**Issue by fmassa** [pytorch/pytorch#9530]
Currently, loading checkpoints using `strict=False` can be fairly error-prone. An extra level (for example with a `nn.DataParallel`) could lead to completely missing loading all the weights, and there is no sign that this has happened.

### Proposal: 
A way to enable user-code to verify for possible mistakes would be to add an option to `load_state_dict` to return matched / unmatched / unexpected keys.
This would allow the user to easily inspect if the loading is actually doing what is expected.

Returning those values is only needed if `strict=False`, because in `strict=True` mode an error is raised if the state dict doesn't match exactly.

A possible API:
```python
def load_state_dict(self, state_dict, strict=True, return_unmatched=False):
    ...
    if return_unmatched:
        return matched, unmatched, unexpected
```

Thoughts?

**Comment by apaszke** [comment]
Why would you want to verify the mismatches in your code? If you use `strict=True` it will report a full list of errors in the exception message.

**Comment by soumith** [comment]
the use-case is to do `strict=False` and at the model level, handle the unmatched / unexpected keys with custom logic. For a large project like detectron that will last a few years, it makes sense for them to handle "old" state dicts or "partial" state dicts with their own logic.

**Comment by bhushan23** [comment]
regarding the API, why not always return matched, unmatched, unexpected set?
If what we return is depended on input parameter, it will be pain for the developers to port to new behavior.
Or 
we want to go ahead with returning only if user is interested in to save some time?

**Comment by bhushan23** [comment]
Regarding interface I see two options - 
1. Keep interface consistent - Return matched, unmatched, unexpected set always 
     Pros- Consistent interface, Portability of new usage and switching between these modes
     Cons - Existing code needs changes
2. Conditional return (Above proposal)
     Pros- Simplifies sources of user no needing this feature, as if this doesn't exists
     Cons- Inconsistent interface, Portability
I personally think consistently across the interface is always good.
Do we have any reason to go with conditional return?

**Comment by apaszke** [comment]
Ok, if we really think there's value in having this method, then can we avoid adding yet another function to `load_state_dict` (which is already pretty complex)? Maybe we could have `diff_state_dict` that would return you the difference between them?

**Comment by bhushan23** [comment]
I liked Introducing `diff_state_dict` approach
Interface of `load_state_dict` remains consistent and pretty intuitive to use for users

**Comment by fmassa** [comment]
@apaszke I'm not sure that this would solve the problem a 100%.

It's currently very easy for an external user to obtain the `matched` / `unmatched` / `unexpected` tensors given the current behavior of `load_state_dict`.

If we add this functionality in a separate function, say `diff_state_dict`, we should enforce that it is in sync with `load_state_dict`.
If it's in the same function, it's easy to enforce. If it's in a different function, and in the future we change the behavior of `load_state_dict` but forget to update `diff_state_dict`, well, the advantage of having `diff_state_dict` is gone.

**Comment by soumith** [comment]
what @fmassa means is that `load_state_dict` can easily have subtly different behavior than `diff_state_dict`. For example, what happens if there are collisions and `load_state_dict` is resolving them in one way, while `diff_state_dict` resolves them in another way.

I see @fmassa 's point, but I dont think it's that hard to keep consistent. We just have to refactor the `load_state_dict` code to use `diff_state_dict` to infer collisions / duplicates.

**Comment by fmassa** [comment]
Having `load_state_dict` use `diff_state_dict` sounds good to me

**Comment by bhushan23** [comment]
I have updated #9620 with prototype as follows:
1. diff_state_dict creates set for local keys and given state_dict keys
2. Does set operations to find match, unmatched and unexpected keys
3. If strict or return_keys is true then only call diff routine
4. Eliminate strict from _load_from_state_dict
5. If Mismatch in keys then early raise if strict is true
    Otherwise call load routine

**Comment by bhushan23** [comment]
ping @fmassa

**Comment by bhushan23** [comment]
@soumith @fmassa 
as per discussion with @apaszke 
we don't see any solid reason to do this now.
We updating API with this change and should have propose reasoning on why is this important.

strict = True notifies about missing and unexpected keys correctly.

**Comment by fmassa** [comment]
After a lot of discussion, we decided not to move on with this proposal.
